datafusion/src/physical_plan/mod.rs

```diff
@@ -509,6 +540,43 @@ pub trait Accumulator: Send + Sync + Debug {
     fn evaluate(&self) -> Result<ScalarValue>;
 }
 
+/// A window accumulator represents a stateful object that lives throughout the evaluation of multiple
+/// rows and generically accumulates values.
+///
+/// An accumulator knows how to:
+/// * update its state from inputs via `update`
+/// * convert its internal state to a vector of scalar values
+/// * update its state from multiple accumulators' states via `merge`
+/// * compute the final value from its internal state via `evaluate`
+pub trait WindowAccumulator: Send + Sync + Debug {
+    /// scans the accumulator's state from a vector of scalars, similar to Accumulator it also
+    /// optionally generates values.
+    fn scan(&mut self, values: &[ScalarValue]) -> Result<Option<ScalarValue>>;
```

Comment:
It would help me if you could explain where the "window" for the window function appears in this trait. I assume you already have a design in mind, so I figured I would ask here

I am thinking about a query like the following

```
select sum(value) OVER (ROWS 5 PRECEDING) FROM ....
```

I think in this case, you end up with 10 aggregate values from 10 different windows, in the the following manner:

```
                                                     
                  1 2 3 4 5 6 7 8 9       input      
                                                     
    window 1      ─                                  
    window 2      ───                                
    window 3      ─────                              
    window 4      ───────                            
    window 5      ─────────                          
    window 6        ─────────                        
    window 7          ─────────                      
    window 8            ─────────                    
    window 9              ─────────                  
```

I would have expected the `WindowAccumulator` to have functions like 

```rust
/// Add a new row to the current window
fn new_row_in_window(ScalarValue);

/// remove a row from the current window
fn remove_row_from_window(ScalarValue);

/// The current value of this function for the given window
fn current_value(ScalarValue);
```

Or possibly something like 

```rust
evaluate(window: &[ArrayRef]) -> ScalarValue
```

Reply:
@alamb good question!

The window word in this trait is purely indicating the fact that window functions will use this. it can be of a better name but...

for a design, there are two complications:
1. multiple window functions, each having different window frames, can be scanning batches at the same time, so i'd want each to create its own window accumulator, remembering to push_back, and remove front, on their own; this trait would not put that into the API, it just _scans_
2. specifically for window that peeks ahead, because batches arrive in async stream, it is not feasible to _peek_, so my plan is to allow them to optionally execute a one time shift upon finishing up

Due to 1 and 2, a best possible state vector for window accumulator is possibly `VecDequeue`. And the name `scan` is because accumulators iterate the list and either optionally produce one value at a time (`max` with order by), or exactly one accumulated value upon finish scanning (`max` with empty over), but not both

Reply:
I see -- looking at how these functions are used for `nth_value` [here](https://github.com/apache/arrow-datafusion/pull/413/files#diff-d8b818bb6ef18a0609ad98574fed66ad4b9ea79338e89d184db7dd5b0218f19aR205) helped me get a better sense for how these traits work

I think as you go through the implementation, adding some additional details to help future implementers of this trait. For example, when it is ok to return values from `scan()` or `scan_batch()` and what the expected total number of rows produced?

However having several examples of of implemented window functions I think will help too so no need to change anything more at this time.